cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,817 @@
+    static constexpr unsigned long long cell_num_per_tile_dim = CNumPerDim;
+    //! Number of cells (total) inside each tile
+    static constexpr unsigned long long cell_num_per_tile = CNumPerTile;
+    //! Types
+    using key_type = Key;                        // tile hash key
+    using value_type = Value;                    // tile No.
+    static constexpr HashTypes hash_type = Hash; // hash table
+    using bis_Type = BlockIndexSpace<ExecutionSpace, N, cell_bits_per_tile_dim,
+                                     cell_num_per_tile_dim, cell_num_per_tile,
+                                     hash_type, key_type, value_type>; // itself
+
+    /*!
+      \brief Constructor
+      \param size_x The size of the block (MPI rank) in dim-x (Unit: tile)
+      \param size_y The size of the block (MPI rank) in dim-y (Unit: tile)
+      \param size_z The size of the block (MPI rank) in dim-z (Unit: tile)
+      \param capacity Expected capacity of the allocator to store the tiles
+      \param rehash_factor The factor multiplies to the capacity for resizing,
+      when tile nums exceed the capcity
+    */
+    BlockIndexSpace( const int size_x, const int size_y, const int size_z,
+                     const value_type capacity, const float rehash_factor )
+        : _tile_table_info( "hash_table_info" )
+        , _tile_inserted_label( "insert_label", size_x * size_y * size_z )
+        , _tile_table( size_x * size_y * size_z )
+        , _op_ijk2key( size_x, size_y, size_z )
+        , _op_key2ijk( size_x, size_y, size_z )
+    {
+        // hash table related init
+        _tile_table.clear();
+        auto tile_table_info_mirror =
+            Kokkos::create_mirror_view( Kokkos::HostSpace(), _tile_table_info );
+        tile_table_info_mirror( 0 ) = 0;
+        tile_table_info_mirror( 1 ) = capacity;
+        Kokkos::deep_copy( _tile_table_info, tile_table_info_mirror );
+
+        // size related init
+        _block_size[0] = size_x;
+        _block_size[1] = size_y;
+        _block_size[2] = size_z;
+        int total_size = size_x * size_y * size_z;
+        _rehash_coeff[0] = rehash_factor;
+
+        // insert status related init
+        auto tile_inserted_label_mirror = Kokkos::create_mirror_view(
+            Kokkos::HostSpace(), _tile_inserted_label );
+        for ( int i = 0; i < total_size; i++ )
+        {
+            tile_inserted_label_mirror( i ) = 0;
+        }
+        Kokkos::deep_copy( _tile_inserted_label, tile_inserted_label_mirror );
+    }
+
+    /*!
+      \brief Valid tile number inside current block (MPI rank)
+    */
+    KOKKOS_INLINE_FUNCTION
+    value_type validTileNum() const { return _tile_table_info( 0 ); }
+
+    /*!
+      \brief Insert a tile into the hash table
+      \param tile_i Tile Id in dim-x
+      \param tile_j Tile Id in dim-y
+      \param tile_k Tile Id in dim-z
+    */
+    KOKKOS_INLINE_FUNCTION
+    void insert( int tile_i, int tile_j, int tile_k ) const
```

Comment:
Here is another way to implement this with only one atomic and without the `_tile_inserted_label_view` but instead just a global `value_type _num_tile` counter!
```
// Get the tile key
key_type tile_key = ijk2key( tile_i, tile_j, tile_k );

// Try to insert the tile key into the map. Give it a dummy value for now.
auto insert_result = _tile_table.insert( tile_key, 0 );

// If the tile key was actually inserted, atomically increment the counter so we know which 
// local id the tile has. Note that this atomic should be pretty scalable because 
// only threads that actually did a successful insert will call it so only one particle 
// in a tile will actually hit this call, substantially reducing contention.
if ( !insert_result.existing() )
    _tile_table.value_at( insert_result.index() ) = Kokkos::atomic_fetch_add( &_num_tile, 1 );
```

Note that `_num_tile` is a class member of type `value_type` that must be initialized to zero before any `insert` sequence. I think this is more scalable because it eliminates the storage of `_tile_inserted_label_view` completely, only has one atomic, and uses the thread-scalable insert in the unordered map instead to decide if the key should go into the table. 

I think this approach also lends itself to better OpenMP thread scaling and potentially a use for duplication in the future.

Reply:
One more thing here - let’s have insert() return the inserted tile index and also have the insert function in the sparse space return the cell index - we may want to use it upon insertion